packages/eslint-plugin/src/rules/no-unsafe-type-assertion.ts

```diff
@@ -0,0 +1,62 @@
+import type { TSESTree } from '@typescript-eslint/utils';
+
+import {
+  createRule,
+  getConstrainedTypeAtLocation,
+  getParserServices,
+} from '../util';
+
+export default createRule({
+  name: 'no-unsafe-type-assertion',
+  meta: {
+    type: 'problem',
+    docs: {
+      description: 'Disallow type assertions that widen a type',
+      requiresTypeChecking: true,
+    },
+    messages: {
+      unsafeTypeAssertion:
+        'Unsafe type assertion: type `{{type}}` is not assignable to type `{{asserted}}`',
```

Comment:
Thinking on the long error messages: we haven't had to really solve this problem before, which is why we don't have good helpers. Most of our error messages either don't mention types or only refer to small names such as identifiers.

For this rule, I think we can generally assume that the explicitly written type should be simple enough to report with. What do you think about this?

> Unsafe type assertion: type '{{type}}' is more narrow than the original type.

I don't love "original type" but 🤷 can't think of anything nicer...